src/transport/async_channel.rs

```diff
@@ -0,0 +1,364 @@
+use futures_sink::Sink;
+
+use crate::{
+    transport::{Connection, ConnectionErrors, LocalAddr, ServerEndpoint},
+    RpcMessage, Service,
+};
+use core::fmt;
+use std::{error, fmt::Display, marker::PhantomData, pin::Pin, result, sync::Arc, task::Poll};
+use tokio::sync::{mpsc, Mutex};
+
+use super::ConnectionCommon;
+
+/// Error when receiving from a channel
+///
+/// This type has zero inhabitants, so it is always safe to unwrap a result with this error type.
+#[derive(Debug)]
+pub enum RecvError {}
+
+impl fmt::Display for RecvError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(self, f)
+    }
+}
+
+/// Sink for memory channels
+pub struct SendSink<T: RpcMessage>(pub(crate) tokio_util::sync::PollSender<T>);
+
+impl<T: RpcMessage> fmt::Debug for SendSink<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SendSink").finish()
+    }
+}
+
+impl<T: RpcMessage> Sink<T> for SendSink<T> {
+    type Error = self::SendError;
+
+    fn poll_ready(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        Pin::new(&mut self.0)
+            .poll_ready(cx)
+            .map_err(|_| SendError::ReceiverDropped)
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        Pin::new(&mut self.0)
+            .start_send(item)
+            .map_err(|_| SendError::ReceiverDropped)
+    }
+
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        Pin::new(&mut self.0)
+            .poll_flush(cx)
+            .map_err(|_| SendError::ReceiverDropped)
+    }
+
+    fn poll_close(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        Pin::new(&mut self.0)
+            .poll_close(cx)
+            .map_err(|_| SendError::ReceiverDropped)
+    }
+}
+
+/// Stream for memory channels
+pub struct RecvStream<T: RpcMessage>(pub(crate) tokio_stream::wrappers::ReceiverStream<T>);
+
+impl<T: RpcMessage> fmt::Debug for RecvStream<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RecvStream").finish()
+    }
+}
+
+impl<T: RpcMessage> Stream for RecvStream<T> {
+    type Item = result::Result<T, self::RecvError>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        match Pin::new(&mut self.0).poll_next(cx) {
+            Poll::Ready(Some(v)) => Poll::Ready(Some(Ok(v))),
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl error::Error for RecvError {}
+
+/// A `tokio::sync::mpsc` based server endpoint.
+///
+/// Created using [connection].
+pub struct MpscServerEndpoint<S: Service> {
```

Comment:
Not quite sure about this, maybe just call it ServerEndpoint?

Reply:
yeah all names are up for discussion

Reply:
renamed